Adding AGEMOEA and IGD

IWNMWE opened the pull request:
This PR implements the IGD score and AGEMOE optimizer. 

Colab Notebook link(IGD test) : https://colab.research.google.com/drive/1fHLPFbOZCmK4Dfwqsw5Ey08lqwRDp9QK?usp=sharing

zoq commented on include/ensmallen_bits/age_moea/agemoea.hpp:
```diff
@@ -0,0 +1,478 @@
+   * @param lowerBound Lower bound of the coordinates of the initial population.
+   * @param upperBound Upper bound of the coordinates of the initial population.
+   */
+  AGEMOEA(const size_t populationSize = 100,
+        const size_t maxGenerations = 2000,
+        const double crossoverProb = 0.6,
+        const double mutationProb = 0.3,
+        const double mutationStrength = 1e-3,
+        const double distributionIndex = 20,
+        const double epsilon = 1e-6,
+        const double eta = 20,
+        const double lowerBound = 0,
+        const double upperBound = 1);
```
See comment above.

zoq commented on include/ensmallen_bits/age_moea/agemoea.hpp:
```diff
@@ -0,0 +1,478 @@
+        const double mutationStrength = 1e-3,
+        const double distributionIndex = 20,
+        const double epsilon = 1e-6,
+        const double eta = 20,
```
`eta` is missing in the parameter description.

zoq commented on include/ensmallen_bits/age_moea/agemoea.hpp:
```diff
@@ -0,0 +1,478 @@
+  //! The tolerance for termination.
+  double epsilon;
+
+  double eta;
```
`eta` is missing in the parameter description.

zoq commented on include/ensmallen_bits/age_moea/agemoea_impl.hpp:
```diff
@@ -0,0 +1,680 @@
+    // Initialize the children to the respective parents.
+    MatType childA = population[indexA], childB = population[indexB];
+
+    if(arma::randu() <= crossoverProb)
+    Crossover(childA, childB, population[indexA], population[indexB]);
```
```suggestion
    if (arma::randu() <= crossoverProb)
      Crossover(childA, childB, population[indexA], population[indexB]);
```

zoq commented on include/ensmallen_bits/age_moea/agemoea_impl.hpp:
```diff
@@ -0,0 +1,680 @@
+    parents.slice(1) = parentB;
+    MatType& current_min =  arma::min(parents, 2);
+    MatType& current_max =  arma::max(parents, 2);
+
```
```suggestion
```

Remove the extra empty line here.

zoq commented on include/ensmallen_bits/age_moea/agemoea_impl.hpp:
```diff
@@ -0,0 +1,680 @@
+    MatType& current_diff = current_max - current_min;
+    current_diff.clamp(1e-10, arma::datum::inf);
+
+    MatType beta1 = 1 + 2.0 * (current_min - lowerBound) / current_diff;
```
Do you think you can add some comments to a few lines here, or at the beginning?

zoq commented on include/ensmallen_bits/age_moea/agemoea_impl.hpp:
```diff
@@ -0,0 +1,680 @@
+                                          const std::vector<size_t>& front,
+                                          size_t dimension)
+{
+  arma::mat final(front.size(),front.size(),arma::fill::zeros);
```
This applies to some other lines as well, but ideally we avoid using `arma::mat` and use `MatType` throughout the code. That way, we could run this with half precision i.e. `arma::fmat` as well.

zoq commented:
Sounds good to me, thanks for the update.

IWNMWE commented:
Hey I have validated its logic with some of the existing problems in mlpack (ZDT suite) and will now proceed to do it with Maf and will put up the PR for the maf problems withing this week. If possible can you review #397 @zoq . Please feel free to also go through the code changes in this PR for any other issues.

zoq commented:
Approved #397 and it should be merged in the next 24 hours.

zoq commented:
Let me know if this is ready for review.

IWNMWE commented:
@zoq Can you please go through this PR and suggest any possible changes.

zoq commented on include/ensmallen_bits/utility/indicators/igd.hpp:
```diff
@@ -0,0 +1,92 @@
+/**
+ * @file igd_plus.hpp
```
Either we adjust the filename or adjust the `@file` here.

zoq commented on tests/agemoea_test.cpp:
```diff
@@ -0,0 +1,332 @@
+{
+  return !(value < (low - roundoff)) && !((high + roundoff) < value);
+}
+
```
```suggestion
```

Remove the extra empty line.

zoq commented on tests/agemoea_test.cpp:
```diff
@@ -0,0 +1,332 @@
+
+      if (!IsInBounds<double>(valX, expectedLowerBound, expectedUpperBound, 0.3) ||
+          !IsInBounds<double>(valY, expectedLowerBound, expectedUpperBound, 0.3) ||
+         !IsInBounds<double>(valZ, expectedLowerBound, expectedUpperBound, 0.3))
```
```suggestion
          !IsInBounds<double>(valZ, expectedLowerBound, expectedUpperBound, 0.3))
```

to align with the other parameters.

zoq commented on include/ensmallen_bits/age_moea/agemoea_impl.hpp:
```diff
@@ -0,0 +1,783 @@
+
+  for (size_t generation = 1; generation <= maxGenerations && !terminate; generation++)
+  {
+
```
```suggestion
```

Remove the extra empty line.

zoq commented on include/ensmallen_bits/age_moea/agemoea_impl.hpp:
```diff
@@ -0,0 +1,783 @@
+  // Calculations for the other fronts.
+  else
+  {
+
```
```suggestion
```

Remove the extra empty line.

zoq reviewed (commented):
Just some really minor style comments, the implementation looks great to me.

zoq reviewed (commented):
Can you run the test case with different random seeds `-rng-seed "time"` and see how often it fails? The Windows build currently fails in the test run, so perhaps we have to adjust the tolerance.

rcurtin commented:
@IWNMWE can you be sure to add documentation to `doc/` for the new technique?  Thanks :+1:

IWNMWE commented:
> Can you run the test case with different random seeds `-rng-seed "time"` and see how often it fails? The Windows build currently fails in the test run, so perhaps we have to adjust the tolerance.

Hey with latest parameter update my tests seem to be running fine on the windows build will try again after the documentation changes mentioned by @rcurtin

zoq commented:
Yeah, some of the tests case fail, due to a bad rng seed, we are trying to minimize the problem by finding some robust hyper parameters.

IWNMWE commented:
@zoq although it was running properly in my system for different seeds I added another try to make it more robust

zoq commented:
> @zoq although it was running properly in my system for different seeds I added another try to make it more robust

Great, I have seen this on my local system as well, it passes all the tests but then on another system it fails.

IWNMWE commented:
Looks like some other test is failing should I make a small commit and re run the pipeline ?

zoq commented:
@mlpack-jenkins test this please

zoq commented:
> Looks like some other test is failing should I make a small commit and re run the pipeline ?

Triggered it again

zoq reviewed (approved):
No further comments from my side, looks great to me.

rcurtin commented on include/ensmallen_bits/age_moea/agemoea.hpp:
```diff
@@ -0,0 +1,489 @@
+   * @return true if the first candidate is preferred, otherwise, false.
+   */
+  template<typename MatType>
+  bool SurvivalScoreOperator(size_t idxP,
+                             size_t idxQ,
+                             const std::vector<size_t>& ranks,
+                             const std::vector<typename MatType::elem_type>& survivalScore);
```
```suggestion
  bool SurvivalScoreOperator(
      size_t idxP,
      size_t idxQ,
      const std::vector<size_t>& ranks,
      const std::vector<typename MatType::elem_type>& survivalScore);
```
Same here, just to keep lines from being too long.

rcurtin commented on include/ensmallen_bits/age_moea/agemoea.hpp:
```diff
@@ -0,0 +1,489 @@
+  */
+ template <typename MatType>
+ void NormalizeFront(
+    std::vector<arma::Col<typename MatType::elem_type> >& calculatedObjectives,
+                    arma::Col<typename MatType::elem_type>& normalization,
+                    const std::vector<size_t>& front,
+                    const arma::Row<size_t>& extreme);
```
```suggestion
     std::vector<arma::Col<typename MatType::elem_type>>& calculatedObjectives,
     arma::Col<typename MatType::elem_type>& normalization,
     const std::vector<size_t>& front,
     const arma::Row<size_t>& extreme);
```

rcurtin commented on include/ensmallen_bits/age_moea/agemoea.hpp:
```diff
@@ -0,0 +1,489 @@
+ double GetGeometry(
+      std::vector<arma::Col<typename MatType::elem_type> >& calculatedObjectives,
+                    const std::vector<size_t>& front,
+                    const arma::Row<size_t>& extreme);
```
Can you go through these function signatures and make sure they match the mlpack style guide?  I did the one above, but it seems like most of these are formatted a little off. :+1:

rcurtin commented on include/ensmallen_bits/age_moea/agemoea_impl.hpp:
```diff
@@ -0,0 +1,781 @@
+                lowerBound, upperBound);
+
+    Mutate(childA, 1.0 / static_cast<double>(numVariables),
+          lowerBound, upperBound);
```
Wrapped lines should be indented by 4 spaces (looks like this is 6?)

rcurtin commented on include/ensmallen_bits/age_moea/agemoea_impl.hpp:
```diff
@@ -0,0 +1,781 @@
+                             const MatType& lowerBound,
+                             const MatType& upperBound)
+{
+    //! Generates a child from two parent individuals
```
It seems like the indentation in alignment both in this function definition and the function body itself are off.

rcurtin commented on include/ensmallen_bits/age_moea/agemoea_impl.hpp:
```diff
@@ -0,0 +1,781 @@
+  std::set<size_t>::iterator it;
+  for (it = selected.begin(); it != selected.end(); it++)
+  {
+    if(*it == S){ continue; }
```
```suggestion
    if (*it == S)
      continue;
```
Here is a link to the style guide, if you don't mind going through the whole code, I haven't checked everything in this file (but skimming through I can see there are a lot of places where it doesn't look quite right): https://github.com/mlpack/mlpack/wiki/DesignGuidelines#style-guidelines

rcurtin commented on include/ensmallen_bits/age_moea/agemoea_impl.hpp:
```diff
@@ -0,0 +1,781 @@
+
+} // namespace ens
+
+#endif
```
This is a minor issue too, but if you can make sure files have newlines at the end of them, I would appreciate it 👍

rcurtin commented on include/ensmallen_bits/utility/indicators/igd.hpp:
```diff
@@ -0,0 +1,92 @@
+ * year={2004},
+ * organization={Springer}
+ * }
+ * @endcode
```
It might be nice to add a little documentation mentioning what the purpose of the indicator is (I guess that it is used by AGEMOEA?).  If this is only used by AGEMOEA, I might suggest putting it in the `age_moea/` directory.

IWNMWE replied:
It is a seperate indicator with a simple change in the already implemented igd_plus score which is used in various papers as the main evaluation metric.

rcurtin replied:
I don't see anything in this PR about the "igd_plus score"; can you clarify what you mean?  If you mean that this class is a standalone utility to measure the quality of a multiobjective solution, are we using it anywhere or in any tests?  I don't see it actually used in this PR.

Do you think we can add tests for it?  And, it should be mentioned in the markdown documentation in `doc/` somewhere, because if not, a user will never know it exists.  I might suggest in the `function_types.md` documentation for multi-objective functions.  If we add enough classes of this type, it might make sense to create a separate documentation file for that, but that can happen at a later time.

Lastly I might suggest changing the name `indicators` to something more like `scores` or `metrics` or similar.  I've never seen the work "indicator" used in this way; in fact, you even used the word 'metric' in the description of the class, so maybe that is the best name to use for a directory containing these.

IWNMWE replied:
Hey, Sure I will add the tests I had done for verification and the reference implementation for the IGD performance indicator with the suggested style changes.

By the above point I actually meant that IGD is very similar to IGD+ which is already implemented in the library and keeping them both together in the indicators should be more better

Regarding the directory name I would suggest to keep the indicator name since these metrics are famously known as `performance indicators `in [literature](https://www.researchgate.net/publication/346979948_Performance_indicators_in_multiobjective_optimization), but yes people do use the name `metrics` as well for it.

Finally for the documentation I will add these performance indicators description in the `function_types.md` file

rcurtin replied:
Oh wow, I see now that this is not the first thing in the `indicators/` directory.  I didn't know about the `IGDPlus` class; I see there is an `Epsilon` class too.  (Those are both tested, but not documented.)  If you want to keep the indicator name, that's fine with me (I wasn't familiar with the paper).

When you add the performance indicators description, if you don't mind adding a bit for `IGDPlus` and `Epsilon` too, that would be great; otherwise from a user perspective they effectively don't exist. :)  (Alternately, if you just want to open an issue about it, or do it in a separate PR, that's fine with me, but we should definitely remember to do it at some point!)

rcurtin commented on include/ensmallen_bits/age_moea/agemoea.hpp:
```diff
@@ -0,0 +1,489 @@
+ * @endcode
+ *
+ */
+class AGEMOEA
```
If the directory is named `age_moea`, should we name this `AGE_MOEA`?  Or should we change the directory name to `agemoea/`?  I'm just looking for consistency, I don't have an opinion either way 👍

rcurtin reviewed (changes requested):
Hey there @IWNMWE, I had been meaning to take a look at this, and sorry it took so long.  I have a lot of comments but they are all about documentation, organization, and style---maybe that seems tedious, but it's important for the sake of consistency and maintainability (which in turn is super important for bringing in new contributors and keeping the library easy to use and develop!).

(I don't know AGEMOEA so I did not review the mathematical details at all.)

rcurtin commented on doc/optimizers.md:
```diff
@@ -2107,7 +2167,7 @@ size equal to that of the starting population.
 #### Constructors
 
  * `NSGA2()`
- * `NSGA2(`_`populationSize, maxGenerations, crossoverProb, mutationProb, mutationStrength, epsilon, lowerBound, upperBound`_`)`
```
Why remove this parameter from the documentation?  It's still in the code (and the table below).

IWNMWE replied:
Sorry for the mistake reverted it

rcurtin commented on include/ensmallen_bits/agemoea/agemoea.hpp:
```diff
@@ -0,0 +1,496 @@
+  template<typename MatType,
+           typename... ArbitraryFunctionType,
+           typename... CallbackTypes>
+ typename MatType::elem_type Optimize(
+     std::tuple<ArbitraryFunctionType...>& objectives,
+     MatType& iterate,
+     CallbackTypes&&... callbacks);
```
```suggestion
  typename MatType::elem_type Optimize(
      std::tuple<ArbitraryFunctionType...>& objectives,
      MatType& iterate,
      CallbackTypes&&... callbacks);
```

rcurtin commented on include/ensmallen_bits/agemoea/agemoea_impl.hpp:
```diff
@@ -0,0 +1,806 @@
+    vectorizedObjectives.col(i) = calculatedObjectives[front[i]];
+  }
+
+  if(front.size() < numObjectives)
```
```suggestion
  if (front.size() < numObjectives)
```

rcurtin commented on include/ensmallen_bits/agemoea/agemoea_impl.hpp:
```diff
@@ -0,0 +1,806 @@
+  size_t index = arma::index_min(d);
+  double avg = arma::accu(calculatedObjectives[front[index]]) / static_cast<double> (numObjectives); 
+  double p = std::log(numObjectives) / std::log(1.0 / avg);
+  if (p <= 0.1 || std::isnan(p)) {p = 1.0;}
```
```suggestion
  if (p <= 0.1 || std::isnan(p))
    p = 1.0;
```
Seems clearer to just wrap the line to me.

rcurtin commented on include/ensmallen_bits/agemoea/agemoea_impl.hpp:
```diff
@@ -0,0 +1,806 @@
+{
+  typedef typename MatType::elem_type ElemType;
+  
+  if(numObjectives >= front.size())
```
```suggestion
  if (numObjectives >= front.size())
```

rcurtin commented on include/ensmallen_bits/agemoea/agemoea_impl.hpp:
```diff
@@ -0,0 +1,806 @@
+//! Check if a candidate Pareto dominates another candidate.
+template<typename MatType>
+inline bool AGEMOEA::Dominates(
+    std::vector<arma::Col<typename MatType::elem_type> >& calculatedObjectives,
```
Why not just use `ColType` as a template parameter directly in this function and other functions?  It seems very awkward to do it like this because `MatType` is not actually used in these functions (at least I did not see any usage at a quick glance).

IWNMWE replied:
I also felt the same but decided to keep it there since a similar pattern was followed `nsga2 `so I kept `agmoea` similar to it aiming for uniform implementation as `agemoea` is very similar instructure to `nsga2`.

I can do the changes but I recommend doing these changes for all the algorithms together in a later PR.

rcurtin replied:
It's a bit of a strange pattern, personally I think it would be a good idea to change it (it's clearer, and the way it currently is doesn't seem to provide any benefit).  Could be worth changing in NSGA2 too.

IWNMWE replied:
Yes, I think its better if both are done together so I can pick it up after NSGA3 has been Implemented (Using the ColType ?) and then I can work on these two

rcurtin commented on include/ensmallen_bits/agemoea/agemoea_impl.hpp:
```diff
@@ -0,0 +1,806 @@
+  typedef typename MatType::elem_type ElemType;
+
+  // Calculations for the first front.
+  if(fNum == 0){
+
+    if(front.size() < numObjectives)
```
```suggestion
  if (fNum == 0)
  {
    if (front.size() < numObjectives)
```

rcurtin commented on include/ensmallen_bits/agemoea/agemoea_impl.hpp:
```diff
@@ -0,0 +1,806 @@
+  // Calculations for the other fronts.
+  else
+  {
+    for(size_t i = 0; i < front.size(); i++)
```
```suggestion
    for (size_t i = 0; i < front.size(); i++)
```

zoq reviewed (commented):
As an fyi, I have some local style changes, going to open a PR, once this is merged.

rcurtin commented:
Sounds good.  The only comments that matter to me before approval are:

 * [NSGA2 documentation change](https://github.com/mlpack/ensmallen/pull/399#discussion_r1690539090)
 * [indicator documentation](https://github.com/mlpack/ensmallen/pull/399#discussion_r1693168037) (just a plan for the other indicators is fine, so long as we do address it at some point)
 * [template type cleanup](https://github.com/mlpack/ensmallen/pull/399#pullrequestreview-2197907775) but that is not a huge deal if we don't do it

Sorry to hold things up a bit on this PR.

IWNMWE commented:
@rcurtin I have made the requested changes. The documentation for Indicators I have added to `function_types.md` and also fixed the NSGA2 documentation. I have made the requested style fixes as well.As mentioned before I plan on doing the `ColType` changes for NSGA2 and AGEMOEA after I finish NSGA3 implementation (using `ColType`) due time constraint.

rcurtin commented on doc/optimizers.md:
```diff
@@ -501,6 +501,66 @@ optimizer.Optimize(f, coordinates);
+#### See also:
+
+ * [An adaptive evolutionary algorithm based on non-euclidean geometry for many-objective optimization](https://doi.org/10.1145/3321707.3321839)
+ * [Multi-Objective Optimization in Wikipedia](https://en.wikipedia.org/wiki/Multi-objective_optimization)
```
Up to you but it could also be useful to link to the indicators documentation here, or perhaps even use an indicator in the example. :+1:

IWNMWE replied:
I have added a reference front comparision of nsga2 and agemoe using IGDas an example please do check it out

IWNMWE commented on doc/function_types.md:
```diff
@@ -1194,3 +1195,85 @@ including ensmallen:
 ignored.  However, be aware that the code may fail to compile and give more
 confusing and difficult error messages!
+
+# Performance Indicators
```
Ok I will add the introduction and move it to the multi objective function subsection

rcurtin commented on doc/function_types.md:
```diff
@@ -884,13 +884,96 @@ arma::cube bestFront = optimizer.ParetoFront();
+
+Performance indicators in multiobjective optimization provide essential metrics 
+for evaluating solution quality, such as convergence to the Pareto front and solution 
+diversity. These indicators are vital for understanding trade-offs and guiding 
+algorithm selection.
```
```suggestion
diversity.
```
Just a suggestion, that sentence doesn't add anything.

rcurtin commented on doc/function_types.md:
````diff
@@ -884,13 +884,96 @@ arma::cube bestFront = optimizer.ParetoFront();
+</summary>
+
+```c++
+  arma::cube referenceFront(2, 1, 3);
+  double tol = 1e-10;
+  referenceFront.slice(0) = arma::vec{0.01010101, 0.89949622};
+  referenceFront.slice(1) = arma::vec{0.02020202, 0.85786619};
+  referenceFront.slice(2) = arma::vec{0.03030303, 0.82592234};
+  arma::cube front = referenceFront * 1.1;
+  double eps =  Epsilon::Evaluate(front, referenceFront);
````
```suggestion
arma::cube referenceFront(2, 1, 3);
double tol = 1e-10;
referenceFront.slice(0) = arma::vec{0.01010101, 0.89949622};
referenceFront.slice(1) = arma::vec{0.02020202, 0.85786619};
referenceFront.slice(2) = arma::vec{0.03030303, 0.82592234};
arma::cube front = referenceFront * 1.1;
double eps = Epsilon::Evaluate(front, referenceFront);
```
No need for indentation here.  Also, it could be nice to add some comments indicating what the expected output is, or what you might do with `eps`.  Still, super nice to have an example either way 👍

rcurtin commented on doc/function_types.md:
````diff
@@ -884,13 +884,96 @@ arma::cube bestFront = optimizer.ParetoFront();
+</summary>
+
+```c++
+  arma::cube referenceFront(2, 1, 3);
+  double tol = 1e-10;
+  referenceFront.slice(0) = arma::vec{0.01010101, 0.89949622};
+  referenceFront.slice(1) = arma::vec{0.02020202, 0.85786619};
+  referenceFront.slice(2) = arma::vec{0.03030303, 0.82592234};
+  arma::cube front = referenceFront * 1.1;
+  // The third parameter is the power constant in the distance formula.
+  double igd =  IGD::Evaluate(front, referenceFront, 1);
````
```suggestion
arma::cube referenceFront(2, 1, 3);
double tol = 1e-10;
referenceFront.slice(0) = arma::vec{0.01010101, 0.89949622};
referenceFront.slice(1) = arma::vec{0.02020202, 0.85786619};
referenceFront.slice(2) = arma::vec{0.03030303, 0.82592234};
arma::cube front = referenceFront * 1.1;
// The third parameter is the power constant in the distance formula.
double igd = IGD::Evaluate(front, referenceFront, 1);
```

rcurtin commented on doc/function_types.md:
````diff
@@ -884,13 +884,96 @@ arma::cube bestFront = optimizer.ParetoFront();
+</summary>
+
+```c++
+  arma::cube referenceFront(2, 1, 3);
+  double tol = 1e-10;
+  referenceFront.slice(0) = arma::vec{0.01010101, 0.89949622};
+  referenceFront.slice(1) = arma::vec{0.02020202, 0.85786619};
+  referenceFront.slice(2) = arma::vec{0.03030303, 0.82592234};
+  arma::cube front = referenceFront * 1.1;
+  double igdPlus =  IGDPlus::Evaluate(front, referenceFront);
````
```suggestion
arma::cube referenceFront(2, 1, 3);
double tol = 1e-10;
referenceFront.slice(0) = arma::vec{0.01010101, 0.89949622};
referenceFront.slice(1) = arma::vec{0.02020202, 0.85786619};
referenceFront.slice(2) = arma::vec{0.03030303, 0.82592234};
arma::cube front = referenceFront * 1.1;
double igdPlus = IGDPlus::Evaluate(front, referenceFront);
```

rcurtin commented on doc/optimizers.md:
```diff
@@ -501,6 +501,97 @@ optimizer.Optimize(f, coordinates);
+ 
+arma::cube NSGAFront = opt2.ParetoFront();
+// Get the IGD score for NSGA front using AGEMOEA as reference.
+double igd =  IGD::Evaluate(NSGAFront, bestFront, 1);
```
```suggestion
double igd = IGD::Evaluate(NSGAFront, bestFront, 1);
```

rcurtin commented on doc/optimizers.md:
```diff
@@ -501,6 +501,97 @@ optimizer.Optimize(f, coordinates);
+
+ * [An adaptive evolutionary algorithm based on non-euclidean geometry for many-objective optimization](https://doi.org/10.1145/3321707.3321839)
+ * [Multi-Objective Optimization in Wikipedia](https://en.wikipedia.org/wiki/Multi-objective_optimization)
+ * [Performance Indicators](#performance-indicators)
```
Could be nice to also add this link to NSGA and MOEA documentation too---but up to you (and doesn't need to happen in this PR).

rcurtin reviewed (approved):
Thanks for addressing everything @IWNMWE!  Sorry that there is so much trickiness and pickiness over documentation; it's what users see, so ensuring that everything is consistent everywhere is important.  I left a few more suggestions, but they are only suggestions---don't feel obligated to take them.  Nothing else from my side, thanks again for the hard work on this. 🚀

IWNMWE commented:
I have made all the required changes except the ColType which I can do the end of august for NSGA2 and AGEMOEA. Other than that I think this PR is good to be merged. Please let me know if there are other changes I have to make @zoq

zoq reviewed (commented):
Thanks for putting this together, I will open another PR to fix some style issues.
